stackhpc_ipa_hardware_managers/system_nic.py

```diff
@@ -0,0 +1,367 @@
+        is checked against the expected version specified in the
+        nic_firmware property's firmware_version field.
+
+        """
+
+        if _is_nic_verification_disabled(node):
+            LOG.warning('NIC firmware version verification has been disabled.')
+            return False
+
+        if "extra" not in node or "nic_firmware" not in node["extra"]:
+            raise errors.CleaningError(
+                "Expected property 'nic_firmware' not found. " +
+                _HELP_MSG_EXAMPLE)
+
+        firmware_matchers = node["extra"]["nic_firmware"]
+
+        if not isinstance(firmware_matchers, list):
+            raise errors.CleaningError(
+                "The property 'nic_firmware' should be a list. " +
+                _HELP_MSG_EXAMPLE
+            )
+
+        lookup_table = _get_pci_id_lookup_table(firmware_matchers)
+        successes, failures = self.process_expected_versions(lookup_table)
+
+        error_msgs = []
+        for interface, result in failures.iteritems():
+            msg = (
+                "Firmware version mismatch for card: {interface}. The "
+                "expected version was: {expected_version}, "
+                "but the actual version was {actual_version}. "
+                "The matcher that failed was {matcher}"
+            ).format(
+                interface=interface,
+                expected_version=result.expected_version,
+                actual_version=result.actual_version,
+                matcher=result.matcher)
+            error_msgs.append(msg)
+
+        if failures:
+            raise errors.CleaningError(
+                "Found {} firmware version mismatches "
+                "when verifying NIC firmware. The errors were: \n"
+                .format(len(error_msgs)) +
+                "\n".join(error_msgs)
+            )
+
+        return successes
+
+    def process_expected_versions(self, pci_matcher_lookup_table):
+        """Processes all matching rules
+
+        where a matching rule is a dict with the following required fields:
+
+        * vendor_id : a numeric pci vendor id
+        * device_id: a numeric pci device id
+        * firmware_version: expected firmware version
+
+        :param pci_matcher_lookup_table: dictionary mapping pci_id to a
+            matching rule
+        :return: tuple with the structure (successes, failures) where,
+            *successes* and *failures* are dictionaries mapping the
+            interface name to a NICFirmwareVerifyResult
+        """
+
+        interfaces = self.get_interface_descriptors()
+        successes = {}
+        failures = {}
+        assert type(self.get_interface_descriptors()) == list
+        for interface in interfaces:
+            actual_version = interface.firmware_version
+            interface_name = interface.name
+            pci_id = interface.pci_id
+            if pci_id not in pci_matcher_lookup_table:
+                LOG.warning(_MISSING_MATCHER_RULE
```

Comment:
In some ways I still wonder if we should fail here, because if you made a mistake with the ```pci_id``` it would look like NIC firmware version checking was working when it actually wasn't. Since you can't easily see the warning you could imagine people being oblivious to it.

Perhaps the SE linux style mode flag that you suggested would be useful here so that the plugin could be configured to fail cleaning on warnings. 

I suppose if the additional cleaning plugin for verifying device trees were present we could avoid this issue, if it used the config from this plugin.

@markgoddard, @oneswig  - any more thoughts?

Reply:
I believe the device tree checker should run with a higher priority and would therefore detect this case before it is an issue?

Reply:
Yeah - i think so long as we *ensure* the same ids are being used there's nothing to worry about.

Reply:
I think this is ok, as long as we catch the case (and fail) where a specified matcher does not match any NICs.

Reply:
This is now implemented.